public/dorequest.php

```diff
@@ -257,14 +313,26 @@ function DoRequestError(string $error, ?int $status = 200, ?string $code = null)
     case "awardachievement":
         $achIDToAward = (int) request()->input('a', 0);
         $hardcore = (bool) request()->input('h', 0);
+        $validationHash = request()->input('v');
 
         /**
          * Prefer later values, i.e. allow AddEarnedAchievementJSON to overwrite the 'success' key
          * TODO refactor to optimistic update without unlock in place. what are the returned values used for?
          */
         $response = array_merge($response, unlockAchievement($user, $achIDToAward, $hardcore));
 
-        if (Achievement::where('ID', $achIDToAward)->exists()) {
+        $foundAchievement = Achievement::where('ID', $achIDToAward)->first();
+        if ($foundAchievement !== null) {
+            if (
+                $delegateTo
+                && strcasecmp(
+                    $validationHash,
+                    $foundAchievement->unlockValidationHash($foundDelegateToUser, (int) $hardcore)
+                ) !== 0
+            ) {
+                return DoRequestError('Access denied.', 403, 'access_denied');
+            }
```

Comment:
You should move this above line 322. `unlockAchievement` has some side-effects if it thinks the achievement will be unlocked.

Reply:
Good point, moved. To retain the same error reporting, I needed to add an `else` block:

```php
        } else {
            $response['Error'] = "Data not found for achievement {$achIDToAward}";
            $response['Success'] = false;
        }
```